bokeh/server/contexts.py

```diff
@@ -291,17 +301,27 @@ def should_discard_ignoring_block(session):
 #-----------------------------------------------------------------------------
 
 class _RequestProxy(object):
-    def __init__(self, request):
+    def __init__(self, request, cookies=None, headers=None):
         self._request = request
 
         arguments = dict(request.arguments)
         if 'bokeh-session-id' in arguments: del arguments['bokeh-session-id']
         self._arguments = arguments
+        self._cookies = cookies if cookies else dict(request.cookies)
+        self._headers = headers if headers else dict(request.headers)
 
     @property
     def arguments(self):
         return self._arguments
 
+    @property
+    def cookies(self):
+        return self._cookies
+
+    @property
+    def headers(self):
+        return self._headers
```

Comment:
Let me know if you think this makes sense, basically this will override the `curdoc().session_context.request.headers` and ``curdoc().session_context.request.cookies` if the token provides them.

Reply:
could set to empty dicts if they are not available so its always the same type

Reply:
That should already be the case, if they're not provided by the token then it'll just grab them off the request: 

```python
self._cookies = cookies if cookies else dict(request.cookies)
self._headers = headers if headers else dict(request.headers)
```